src/lib/npmCommand.ts

```diff
@@ -160,11 +156,73 @@ export class NpmModule {
   }
 
   public show(registry: string): NpmShowResults {
-    return NpmCommand.runNpmCmd(`show ${this.module}@${this.version}`, { registry, cliRoot: this.cliRoot });
+    const showCmd = NpmCommand.runNpmCmd(`show ${this.module}@${this.version} --json`, {
+      registry,
+      cliRoot: this.cliRoot,
+    });
+
+    // `npm show` doesn't return exit code 1 when it fails to get a specific package version
+    // If `stdout` is empty then no info was found in the registry.
+    if (showCmd.stdout === '') {
+      throw new SfdxError(`Failed to find ${this.module}@${this.version} in the registry`, 'NpmError');
+    }
+
+    try {
+      return JSON.parse(showCmd.stdout) as NpmShowResults;
+    } catch (error) {
+      throw new SfdxError(error, 'ShellParseError');
+    }
   }
 
   public pack(registry: string, options?: shelljs.ExecOptions): void {
-    NpmCommand.runNpmCmd(`pack ${this.module}@${this.version}`, { ...options, registry, cliRoot: this.cliRoot });
+    try {
+      NpmCommand.runNpmCmd(`pack ${this.module}@${this.version}`, {
+        ...options,
+        registry,
+        cliRoot: this.cliRoot,
+      });
+    } catch (err) {
+      const sfdxErr = SfdxError.wrap(err);
+      throw new SfdxError(`Failed to fetch tarball from the registry: \n${sfdxErr.message}`, 'NpmError');
+    }
     return;
   }
+
+  public async fetchTarbarll(registry: string, options?: shelljs.ExecOptions): Promise<void> {
+    await this.pollForAvailability(() => {
+      this.pack(registry, options);
+    });
+    this.pack(registry, options);
+  }
+
+  public async pollForAvailability(checkFn: () => void): Promise<void> {
+    const ux = await UX.create();
+    const isNonTTY = process.env.CI !== undefined || process.env.CIRCLECI !== undefined;
+    let found = false;
+    let attempts = 0;
+    const maxAttempts = 300;
+
+    const start = isNonTTY ? (msg: string): UX => ux.log(msg) : (msg: string): void => ux.startSpinner(msg);
+    const update = isNonTTY ? (msg: string): UX => ux.log(msg) : (msg: string): void => ux.setSpinnerStatus(msg);
+    const stop = isNonTTY ? (msg: string): UX => ux.log(msg) : (msg: string): void => ux.stopSpinner(msg);
+
+    start('Polling for new version(s) to become available on npm');
+    while (!found && attempts < maxAttempts) {
```

Comment:
there's a polling client in sfdx-core if you need that much, or the raw (approved) library here: https://github.com/normartin/ts-retry-promise

Reply:
Tried using the polling client in sfdx-core but ended up with the same code wrapped inside of it.
We still need to write the promise around `this.pack` and pass it to sfdx-core.